### PR TITLE
[ValueTracking] Compute known FPClass from dominating condition

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch improves `computeKnownFPClass` by using context-sensitive information from `DomConditionCache`.

See also the comment https://github.com/llvm/llvm-project/pull/80740#discussion_r1479328484.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -4213,9 +4213,56 @@ llvm::fcmpImpliesClass(CmpInst::Predicate Pred, const Function &F, Value *LHS,
...
+                                        const Instruction *CxtI,
+                                        KnownFPClass &KnownFromContext) {
+  CmpInst::Predicate Pred;
+  Value *LHS;
+  Value *RHS;
+  uint64_t ClassVal = 0;
+  // TODO: handle sign-bit check idiom
+  if (match(Cond, m_FCmp(Pred, m_Value(LHS), m_Value(RHS)))) {
+    const APFloat *CRHS;
+    if (match(RHS, m_APFloat(CRHS))) {
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:4227]: Could move this directly into the initial m_FCmp match; however I do have a patch I was working on to generalize fcmpImpliesClass to handle non-constant cases

**Review by arsenm — COMMENTED** [review]
Could use a few tests where the dominating condition is the fcmp + fabs/fneg case

**Comment by dtcxzyw** [comment]
The regression https://github.com/dtcxzyw/llvm-opt-benchmark/pull/199#discussion_r1480974120 has been fixed.

File: llvm/lib/Analysis/DomConditionCache.cpp [hunk 2]
```diff
@@ -69,6 +69,13 @@ static void findAffectedValues(Value *Cond,
       }
     }
   }
+  // Handle patterns that computeKnownFPClass() support.
+  FCmpInst::Predicate Pred;
+  Value *A;
+  if (match(Cond, m_FCmp(Pred, m_Value(A), m_Constant())))
+    AddAffected(A);
+  if (match(Cond, m_Intrinsic<Intrinsic::is_fpclass>(m_Value(A), m_Constant())))
```
Comment (arsenm) [inline llvm/lib/Analysis/DomConditionCache.cpp:77]: This can be else if

File: llvm/lib/Analysis/DomConditionCache.cpp [hunk 3]
```diff
@@ -69,6 +69,13 @@ static void findAffectedValues(Value *Cond,
       }
     }
   }
+  // Handle patterns that computeKnownFPClass() support.
+  FCmpInst::Predicate Pred;
+  Value *A;
+  if (match(Cond, m_FCmp(Pred, m_Value(A), m_Constant())))
```
Comment (arsenm) [inline llvm/lib/Analysis/DomConditionCache.cpp:75]: else if

**Review by arsenm — COMMENTED** [review]
lgtm with nit

File: llvm/lib/Analysis/DomConditionCache.cpp (hunk 2 above)
Comment (dtcxzyw) [reply llvm/lib/Analysis/DomConditionCache.cpp]: Done.

**Comment by dtcxzyw** [comment]
Ping.

**Review by arsenm — APPROVED** [review]

**Comment by eaeltsin** [comment]
Hi,

This seems to break a couple of our tests, can you please take a look?

[Archive.zip](https://github.com/llvm/llvm-project/files/14302127/Archive.zip) is a source ll file and files obtained
 by `opt -O1` before and after this change.

Thank you!

**Comment by dtcxzyw** [comment]
> Hi,
> 
> This seems to break a couple of our tests, can you please take a look?
> 
> [Archive.zip](https://github.com/llvm/llvm-project/files/14302127/Archive.zip) is a source ll file and files obtained by `opt -O1` before and after this change.
> 
> Thank you!

I will have a look.

**Comment by dtcxzyw** [comment]
Reduced test case: https://alive2.llvm.org/ce/z/YgQSin
I will post a fix later.

**Comment by dtcxzyw** [comment]
@eaeltsin Nice catch! Should be fixed by #81972.

**Pull request merged**